src/webgpu/shader/validation/shader_io/interpolate.spec.ts

```diff
@@ -59,6 +59,47 @@ g.test('type_and_sampling')
     t.expectCompileResult(kValidInterpolationAttributes.has(interpolate), code);
   });
 
+g.test('reversed_type_and_sampling')
+  .desc(`Test that type must come before sampling.`)
+  .params(u =>
+    u
+      .combine('stage', ['vertex', 'fragment'] as const)
+      .combine('io', ['in', 'out'] as const)
+      .combine('use_struct', [true, false] as const)
+      .combine('type', ['', 'flat', 'perspective', 'linear'] as const)
+      .combine('sampling', ['', 'center', 'centroid', 'sample'] as const)
+      .beginSubcases()
+  )
+  .fn(t => {
+    if (t.params.stage === 'vertex' && t.params.use_struct === false) {
+      t.skip('vertex output must include a position builtin, so must use a struct');
+    }
+    if (t.params.type === '' && t.params.sampling === '') {
+      t.skip('both blank is valid');
+    }
+
+    let interpolate = '';
+    if (t.params.type !== '' || t.params.sampling !== '') {
+      interpolate = '@interpolate(';
+      if (t.params.sampling !== '') {
+        interpolate += `${t.params.sampling}`;
+      }
+      if (t.params.type !== '') {
+        interpolate += `, ${t.params.type}`;
+      }
+      interpolate += `)`;
+    }
+    const code = generateShader({
+      attribute: '@location(0)' + interpolate,
+      type: 'f32',
+      stage: t.params.stage,
+      io: t.params.io,
+      use_struct: t.params.use_struct,
+    });
+
+    t.expectCompileResult(false, code);
```

Comment:
I have a slight reservation about validation tests that are expected to always fail. These tests have no way of checking whether they're failing for the right reasons, which means that they have a good chance of becoming stale (e.g. WGSL syntax changes, the test now fails because the syntax is wrong, but we don't notice because fail means pass).

I've been striving to always have control cases for validation tests to avoid this issue, which I think is what the API tests tend to do as well. Is that something that is possible here (and below)?

Reply:
Done. Removed this test and added the invalid values into the above test case.